### PR TITLE
gitUpdater: init

### PR DESCRIPTION
**Pull request by romildo**
###### Motivation for this change

- Add the 'gitUpdater' helper function to update git based packages, using the 'genericUpdater' function.
- Rework argument passing to the `list-git-tags' and 'list-archive-two-level-versions' scripts.
- Replace 'genericUpdater' plus 'list-git-tags' by 'gitUpdater'

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/desktops/mate/default.nix [hunk 1]
```diff
@@ -10,7 +10,7 @@ let
       pkgs.genericUpdater {
         inherit pname version odd-unstable;
         attrPath = "mate.${pname}";
-        versionLister = "${pkgs.common-updater-scripts}/bin/list-archive-two-level-versions ${url}";
```
Comment (jtojnar) [inline pkgs/desktops/mate/default.nix:13]: I wonder if we should have a function that creates this text `pkgs.common-updater-scripts.versionListers.archiveTwoLevelVersions { inherit url; }`.

Comment (romildo) [reply pkgs/desktops/mate/default.nix:13]: Instead I am going to add the function `httpTwoLevelsUpdater`, which calls `genericUpdater` with this argument.

Comment (jtojnar) [reply pkgs/desktops/mate/default.nix:13]: I guess that solves the issue of having to specify the command manually but it is even less compositional than the original variant. I remember wanting to use `gitUpdater` for the community part of gmic-qt but it just could not mesh with the rest:

https://github.com/NixOS/nixpkgs/blob/1822c35423bb208cbdcf6ad75058b38cd5cff88e/pkgs/tools/graphics/gmic-qt/default.nix#L148

**Comment by romildo** [comment]
My idea with the `genericUpdater` is to be able to compose the updater from pieces. It does the following
1. list the available versions
2. find the latest of them
3. update the version in the derivation
4. generate the git commit command

For doing the first part, there are shell scripts to:
1. list available git tags
2. list available file versions from http where the files are organised in version subdirectories (for instance `https://pub.mate-desktop.org/releases/1.17/marco-1.17.0.tar.xz`)

And others can be added. Those are the two cases I am currently using. I want also to add (in a separate PR) another lister for files without the subdirectory for the version (for instance `http://download.enlightenment.org/rel/libs/efl/efl-1.26.2.tar.xz`).

Calling `genericUpdater` for them directly everywhere is becoming too repetitive. Thats why this PR defines two more specialized functions: `gitUpdater` and `httpTwoLevelsUpdater`.


We can make the version lister scripts available through functions in `pkgs.common-updater-scripts.versionListers`, as @jtojnar  propesed, but I would leave it for another PR.

File: pkgs/desktops/mate/default.nix (hunk 1 above)
Comment (romildo) [reply pkgs/desktops/mate/default.nix:13]: The more general function `genericUpdater` is still available.

**Comment by jtojnar** [comment]
I am fine with this as is for now.

By the way, for the committing, you might want to use the experimental `supportedFeatures` option:

https://github.com/NixOS/nixpkgs/commit/c1b05442ffd6cf3cf529cad469bebe8169b156e9

I am adding support for modifying the commit message in https://github.com/NixOS/nixpkgs/pull/161260.

**Pull request merged**